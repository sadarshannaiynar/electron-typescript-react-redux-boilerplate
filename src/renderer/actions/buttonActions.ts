import { IpcMessageEvent, IpcRenderer } from 'electron';
import { Dispatch } from 'redux';
import { action } from 'typesafe-actions';
import { EXEC_LS } from '../../common/constants/commandTypes';
import { DECREMENT_VALUE, EXEC_LS_ERROR, EXEC_LS_SUCCESS, INCREMENT_VALUE, SET_VALUE } from '../constants/actionTypes';

export interface IEventOutput {
  log: string;
}

export const incrementAction = () => action(INCREMENT_VALUE);

export const decrementAction = () => action(DECREMENT_VALUE);

export const setValue = (value: number) => action(SET_VALUE, value);

export const startExecuteLs = () => action(EXEC_LS);

export const successExecuteLs = (output: string) => action(EXEC_LS_SUCCESS, output);

export const errorExecuteLs = (error: string) => action(EXEC_LS_ERROR, error);

export const executeLs = (ipcRenderer: IpcRenderer): any => (
  (dispatch: Dispatch) => {
    ipcRenderer.send(EXEC_LS);
    dispatch(startExecuteLs());

    ipcRenderer.once(`success-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(successExecuteLs(log));
      ipcRenderer.removeAllListeners(`error-${EXEC_LS}`);
    });

    ipcRenderer.once(`error-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(errorExecuteLs(log));
      ipcRenderer.removeAllListeners(`success-${EXEC_LS}`);
    });
  }
);
