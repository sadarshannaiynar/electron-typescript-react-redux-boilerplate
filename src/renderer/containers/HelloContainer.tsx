import { IpcRenderer, ipcRenderer } from 'electron';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { decrementAction, executeLs, incrementAction, setValue } from '../actions/buttonActions';
import { HelloComponent } from '../components/Hello';
import { IRootState } from '../reducers';

export interface IHelloContainerState {
  value: number;
  output: string;
  error: string;
}

export interface IHelloContainerOwnProps {
  store?: any;
}

export interface IHelloContainerDispatch {
  decrement(): void;
  increment(): void;
  setValue(value: number): void;
  executeLs(ipcRenderer: IpcRenderer): any;
}

const mapStateToProps = (state: IRootState): IHelloContainerState => ({
  error: state.buttonReducer.error,
  output: state.buttonReducer.output,
  value: state.buttonReducer.value,
});

const mapDispatchToProps = (dispatch: Dispatch): IHelloContainerDispatch => ({
  decrement: () => {
    dispatch(decrementAction());
  },
  executeLs: () => {
    dispatch(executeLs(ipcRenderer));
  },
  increment: () => {
    dispatch(incrementAction());
  },
  setValue: (value: number) => {
    dispatch(setValue(value));
  },
});

export const HelloContainer = connect<IHelloContainerState, IHelloContainerDispatch, IHelloContainerOwnProps>(
  mapStateToProps, mapDispatchToProps,
)(HelloComponent);
