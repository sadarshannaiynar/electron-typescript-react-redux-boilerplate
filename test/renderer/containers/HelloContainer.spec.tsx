import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import reduxMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { decrementAction, incrementAction, setValue } from '../../../src/renderer/actions/buttonActions';
import { HelloContainer } from '../../../src/renderer/containers/HelloContainer';

interface IWrapperAndMocks {
  mockDispatch: jest.Mock;
  wrapper: ShallowWrapper;
}

const getWrapperAndMocks = (store: MockStoreEnhanced) => {
  const mockDispatch = jest.fn();

  store.dispatch = mockDispatch;

  const wrapper = shallow(<HelloContainer store={store} />);

  return {
    mockDispatch,
    wrapper,
  };
};

describe.only('<HelloContainer /> Tests', () => {
  describe('Rendering Tests', () => {
    test('should match the snapshot', () => {
      const mockStore = reduxMockStore();
      const store = mockStore({
        buttonReducer: {
          error: 'Error',
          output: 'Output',
          value: 1,
        },
      });
      const { wrapper } = getWrapperAndMocks(store);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('State Tests', () => {
    test('should get the value prop from store', () => {
      const mockStore = reduxMockStore();
      const store = mockStore({
        buttonReducer: {
          error: 'Error',
          output: 'Output',
          value: 5,
        },
      });
      const { wrapper } = getWrapperAndMocks(store);
      expect(wrapper.find('HelloComponent').prop('value')).toStrictEqual(5);
      expect(wrapper.find('HelloComponent').prop('output')).toStrictEqual('Output');
      expect(wrapper.find('HelloComponent').prop('error')).toStrictEqual('Error');
    });
  });

  describe('Dispatch Tests', () => {
    const mockStore = reduxMockStore();
    const store = mockStore({
      buttonReducer: {
        error: 'Error',
        output: 'Output',
        value: 5,
      },
    });

    test('should fire dispatch when increment is called', () => {
      const { wrapper, mockDispatch } = getWrapperAndMocks(store);
      mockDispatch.mockClear();
      wrapper.props().increment();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(mockDispatch).toHaveBeenLastCalledWith(incrementAction());
    });

    test('should fire dispatch when decrement is called', () => {
      const { wrapper, mockDispatch } = getWrapperAndMocks(store);
      mockDispatch.mockClear();
      wrapper.props().decrement();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(mockDispatch).toHaveBeenLastCalledWith(decrementAction());
    });

    test('should fire dispatch when setValue is called', () => {
      const { wrapper, mockDispatch } = getWrapperAndMocks(store);
      mockDispatch.mockClear();
      wrapper.props().setValue(5);
      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(mockDispatch).toHaveBeenLastCalledWith(setValue(5));
    });

    test('should fire dispatch when executeLs is called', () => {
      const { wrapper, mockDispatch } = getWrapperAndMocks(store);
      mockDispatch.mockClear();
      wrapper.props().executeLs();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });
});
