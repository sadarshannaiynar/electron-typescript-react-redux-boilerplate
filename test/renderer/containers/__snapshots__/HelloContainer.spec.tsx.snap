// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelloContainer /> Tests Rendering Tests should match the snapshot 1`] = `
<HelloComponent
  decrement={[Function]}
  error="Error"
  executeLs={[Function]}
  increment={[Function]}
  output="Output"
  setValue={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  value={1}
/>
`;
